src/geotop/tensor.h

```diff
@@ -0,0 +1,167 @@
+
+template <class T> class Tensor {
+
+public:
+    /** lower and upper bounds */
+    std::size_t ndh, ndl; // depth
+    std::size_t nrh, nrl; // rows
+    std::size_t nch, ncl; // columns
+
+    std::size_t n_dep;
+    std::size_t n_row;
+    std::size_t n_col;
+
+    /** the actual data */
+    std::unique_ptr<T[]> co; // it has to be written after the lower and upper limits (otherwise: sigfault)
+
+    /** pointer to the first accessible element (needed by an iterator) */
+    T *begin() noexcept { return &co[0]; }
+
+    /** pointer to the one-past the last element (needed by an iterator)*/
+    T *end() noexcept { return &co[(ndh-ndl+1)*(nrh-nrl+1)*(nch-ncl+1)]; }
+
+    /** const pointer to the first element accessible element */
+    const T *begin() const noexcept { return &co[0]; }
+
+    /** const pointer to the one-past the last element */
+    const T *end() const noexcept { return &co[(ndh-ndl+1)*(nrh-nrl+1)*(nch-ncl+1)]; }
+
+
+    /** destructor. default is fine */
+    ~Tensor() = default;
+
+    /** default constructor is deleted */
+    Tensor() = delete;
+
+/**
+   * constructor
+   * @param _nrl,_nrh lower and upper bound for rows
+   * @param _ncl,_nch lower and upper bound for columns
+   * @param _ndl,_ndh lower and upper bound for depth
+   */
+
+    Tensor(const std::size_t _ndh,  const std::size_t _ndl,
+           const std::size_t _nrh, const std::size_t _nrl, const std::size_t _nch,  const std::size_t _ncl):
+            ndh{_ndh}, ndl{_ndl}, nrh{_nrh}, nrl{_nrl}, nch{_nch}, ncl{_ncl},
+            n_dep{ndh-ndl+1}, n_row{nrh-nrl+1}, n_col{nch-ncl+1}, co { new T[(ndh-ndl+1)*(nrh-nrl+1)*(nch-ncl+1)]{} } {}
+    // .................................................................................|-->initialize all elements to 0
+
+    Tensor(const std::size_t d, const std::size_t r, const std::size_t c): Tensor{d,1,r,1,c,1} {}
+
+    /** range-checked access operator */
+    T &at(const std::size_t k, const std::size_t i, const std::size_t j) {
+        GEO_ERROR_IN_RANGE(k, ndl, ndh);
+        GEO_ERROR_IN_RANGE(i, nrl, nrh);
+        GEO_ERROR_IN_RANGE(j, ncl, nch);
+        return (*this)[(i-nrl)*n_col + (j-ncl) + (k-ndl)*(n_row*n_col)];
+    }
+
+    /** range-checked access operator */
+    const T &at(const std::size_t k, const std::size_t i, const std::size_t j) const {
+        GEO_ERROR_IN_RANGE(k, ndl, ndh);
+        GEO_ERROR_IN_RANGE(i, nrl, nrh);
+        GEO_ERROR_IN_RANGE(j, ncl, nch);
+        return (*this)[(i-nrl)*n_col + (j-ncl) + (k-ndl)*(n_row*n_col)];
+    }
+
+    T& operator[](const std::size_t i) noexcept {
+        return co[i];
+    }
+
+    const T& operator[](const std::size_t i) const noexcept {
+        return co[i];
+    }
+
+    /**
+    * access operator. When the code is compiled in debug mode, it performes
+    * a range check. No check is done when the code is compiled in release mode.
+    */
+    T& operator()(const std::size_t k, const std::size_t i, const std::size_t j) {
```

Comment:
please add `noexcept` with the guard `#ifndef NDEBUG` as below

Reply:
done